# User-Agent 부여하기
- User-Agent란? 클라이언트가 서버에 접속하는 플랫폼
    - 현재 우리가 사용하고 있는 User-Agent: 브라우저(크롬, 사파리....)
- 만약 서버에서 크롤링 봇으로 접근하는 것을 차단한다면?
    - ex) 브라우저가 아니면 접근 금지!  하도록 서버가 차단할 수도 있음.
  

**무신사 접속해보기**

In [66]:
import requests
MUSINSA_URL = "https://www.musinsa.com/categories/item/018003"
response = requests.get(MUSINSA_URL)
response.status_code

200

In [67]:
MUSINSA_DETAIL_URL ="https://www.musinsa.com/app/goods/3030422"
response = requests.get(MUSINSA_DETAIL_URL)
response

#무신사는 detail페이지가 크롤링 막혀있음.

<Response [403]>

fake_useragent: User Agent를 랜덤으로 만들어줌.
- 403에러가 뜨면 fake_user넣어줘도 괜찮음

In [68]:
from fake_useragent import UserAgent

headers ={'User-agent':UserAgent().chrome}
headers

{'User-agent': 'Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_6_1; en-US) AppleWebKit/534.3 (KHTML, like Gecko) Chrome/6.0.472.63 Safari/534.3'}

In [69]:
response = requests.get(MUSINSA_DETAIL_URL, headers= headers)
response

<Response [200]>

# Selenium 옵션

In [51]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager

# usergent를 임의로 설정할 수 있는 driver 생성 함수
def get_chrome_driver_useragent():
    chrome_options = webdriver.ChromeOptions()
    
    chrome_options.add_argument("userAgent")
    driver = webdriver.Chrome(
        service=Service(ChromeDriverManager().install()),
        options=chrome_options
    )
    
    return driver

In [52]:
driver = get_chrome_driver_useragent()

In [53]:
driver.execute_script("return navigator.userAgent;")

'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36'

In [54]:
driver.quit()

**Selenium으로 크롤링을 할 때 중요한 옵션**
- Headless: 브라우저를 화면에 띄우지 않고 컴퓨터 내부에서만 돌아가도록
- Windows, Mac같은 GUI(Graphic User Interface) 환경 OS에서는 화면이 뜬다.
- Linux같은 경우는 CLI(Command Line Interface) 환경 OS에서는 명령창만 뜬다.
- 크롬같은 경우는 메모리를 굉장히 많이 잡아먹는 브라우저라서, 메모리를 아끼기 위해 headless옵션을 쓰기도 함.

In [56]:
def get_chrome_driver_headless():
    chrome_options = webdriver.ChromeOptions()
    
    chrome_options.add_argument("headless")
    driver = webdriver.Chrome(
        service=Service(ChromeDriverManager().install()),
        options=chrome_options
    )
    
    return driver

In [57]:
driver = get_chrome_driver_headless()

In [59]:
driver.get("https://www.naver.com")

In [61]:
driver.title

'NAVER'

In [62]:
driver.quit()

* 셀레니움을 이용한 크롤링은 GUI 환경에서 모든 테스트를 마치고 headless 설정을 하여 CLI환경으로 돌린다.